### PR TITLE
Add solution to usaco moortal cowmbat

### PR DESCRIPTION
**Pull request by TruaShamu**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [x] I have tested my code
- [x] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [x] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

### CODE REVIEW

**Comment by TruaShamu** [comment]
Proof that my code works:
java
cpp

File: solutions/usaco-971.mdx [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+**Time Complexity:** $\mathcal{O}(M^2 + N M^2)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+CodeSnip{CPP Short Template}
+
+const int MAXN = 1e5 + 5;
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:21]: remove +5

File: solutions/usaco-971.mdx [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+const int MAXL = 26;
+int freq[MAXN][MAXL];
+
+int count(int start, int end, int let) {
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:25]: short comment explaining this

File: solutions/usaco-971.mdx [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+	}
+	
+	/* 
+	 * dp[i][j] is the minimum cost to make a valid streak
+	 * up to index i, where the last streak has letter j.
+	 */
+	int dp[MAXN][MAXL];
+	/*
+	 * ans[i] is the minimum cost to make a valid streak to index i.
+	 */
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:71]: no multiline

Comment (SansPapyrus683) [reply solutions/usaco-971.mdx:71]: oh wait local arrays are also kinda bad, either
1. make them global (bad)
2. use vectors instead (good)

File: solutions/usaco-971.mdx [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+		// Find a letter 'j' with minimum cost to change all the letters to.
+		ans[idx] = dp[idx][0];
+		for (int j = 1; j < M; j++) {
+			ans[idx] = min(ans[idx], dp[idx][j]);
+		}
+	}
+
+
+	//If (N >= 2K) there is >= 1 unique letter streaks.
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:95]: space after the `//` for comments (other places too)

File: solutions/usaco-971.mdx [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+<JavaSection>
+
+```java
+
+import java.io.*;
+import java.util.*;
+
+public class MoortalCowmbat {
+	public static int MAXN = (int) 1e5 + 5;
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:138]: same thing

File: solutions/usaco-971.mdx [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+			}
+
+		}
+
+		// Output.
+		io.println(ans[N - 1]);
+		io.close();
+
+	}
+	static int count(int start, int end, int let) {
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:237]: whitespace & comment

File: solutions/usaco-971.mdx [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+		}
+
+		/*
+		 * dp[i][j] is the minimum cost to make a valid streak
+		 * up to index i, where the last streak has letter j.
+		 */
+		int[][] dp = new int[MAXN][MAXL];
+		/*
+		 * ans[i] is the minimum cost to make a valid streak to index i.
+		 */
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:182]: no mulltiline

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
so far

File: solutions/usaco-971.mdx (hunk 3 above)
Comment (TruaShamu) [reply solutions/usaco-971.mdx:72]: out of curiosity, why are local arrays bad?
and do i have to change cost, ans, etc to vectors as well?

Comment (SansPapyrus683) [reply solutions/usaco-971.mdx:72]: i would simp for you if you did

**Comment by TruaShamu** [comment]
Ok, done

File: solutions/usaco-971.mdx (hunk 3 above)
Comment (TruaShamu) [reply solutions/usaco-971.mdx]: :thonk:

File: solutions/usaco-971.mdx [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(M^2 + N M^2)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+CodeSnip{CPP Short Template}
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:19]: ...
i'm not mad, just disapointed

File: solutions/usaco-971.mdx [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+	/*
+	 * dp[i][j] is the minimum cost to make a valid streak
+	 * up to index i, where the last streak has letter j.
+	 */
+	vector<vi> dp(N, vi(M, 0x3f));
+	/*
+	 * ans[i] is the minimum cost to make a valid streak to index i.
+	 */
+	vi ans(N, 0x3f);
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:73]: the heck is 0x3f
if you want a really large value just use INT32_MAX or smth

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-971.mdx (hunk 8 above)
Comment (TruaShamu) [reply solutions/usaco-971.mdx:19]: I thought I was supposed to use the codesnippet?
I can remove it if you want, though

Comment (TruaShamu) [reply solutions/usaco-971.mdx]: Alright, the codesnippet is gone, now you can finally rest in peace 🤣

File: solutions/usaco-971.mdx [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+
+// count(i, j, k) is the number of appearances of letter k between index i and j.
+int count(int start, int end, int let) {
+	return freq[end + 1][let] - freq[start][let];
+}
+
+int main() {
+	freopen("cowmbat.in", "r", stdin); // see /general/input-output
+	freopen("cowmbat.out", "w", stdout);
+	int N; int M; int K;
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:34]: commas instead? or each on its own line? not whatever abomination this is

File: solutions/usaco-971.mdx [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+	}
+
+	/*
+	 * dp[i][j] is the minimum cost to make a valid streak
+	 * up to index i, where the last streak has letter j.
+	 */
+	vector<vector<int>> dp(N, vector<int>(M, 0x3f));
+	/*
+	 * ans[i] is the minimum cost to make a valid streak to index i.
+	 */
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:74]: thought i said to remove the multiline?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
i'm going to commit some small stuff, buckle up

File: solutions/usaco-971.mdx [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+		freq[i][S[i - 1] - 'a']++;
+	}
+
+	/*
+	 * dp[i][j] is the minimum cost to make a valid streak
+	 * up to index i, where the last streak has letter j.
+	 */
+	vector<vector<int>> dp(N, vector<int>(M, 0x3f));
+	// ans[i] is the minimum cost to make a valid streak to index i.
+	vector<int> ans(N, 0x3f);
```
Comment (SansPapyrus683) [inline solutions/usaco-971.mdx:75]: what's 0x3f mean

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/usaco-971.mdx [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_cowmbat_gold_dec19.html)
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(M^2 + N M^2)$
+
+<LanguageSection>
+
```
Comment (ryanchou-dev) [inline solutions/usaco-971.mdx:15]:
```suggestion
```

File: solutions/usaco-971.mdx [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+		ans[idx] = dp[idx][0];
+		for (int j = 1; j < M; j++) {
+			ans[idx] = min(ans[idx], dp[idx][j]);
+		}
+
+	}
+	cout << ans[N - 1] << endl;
+}
+```
+
```
Comment (ryanchou-dev) [inline solutions/usaco-971.mdx:126]:
```suggestion
```

File: solutions/usaco-971.mdx [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+</CPPSection>
+
```
Comment (ryanchou-dev) [inline solutions/usaco-971.mdx:128]:
```suggestion
```

File: solutions/usaco-971.mdx [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	// count(i, j, k) is the number of appearances of letter k between index i and j.
+	static int count(int start, int end, int let) {
+		return freq[end + 1][let] - freq[start][let];
+	}
+
+	//CodeSnip{Kattio}
+}
+
```
Comment (ryanchou-dev) [inline solutions/usaco-971.mdx:239]:
```suggestion
```

File: solutions/usaco-971.mdx [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+```
+</JavaSection>
+
```
Comment (ryanchou-dev) [inline solutions/usaco-971.mdx:242]:
```suggestion
```

**Review by ryanchou-dev — APPROVED** [review]
just gonna remove some blank lines, great exp!

**Review by nathangong — APPROVED** [review]

**Review by dongliu0426 — APPROVED** [review]

**Pull request merged**